### DISCUSSION TITLE
How do I make AdonisJS 5 modular and hookable?

### DISCUSSION THREAD

**Discussion by simonjcarr** [adonisjs/core#1355] [Help]
I want to experiment with making a modular application out of AdonisJS 5. I want to provide the ability for third-party developers to write a module that does not require the core application code to be touched.

Many PHP applications have this feature, think WordPress, PrestaShop Drupal etc.

I think the main difference here is that AdonisJS is compiled for production, so modules would have to be added in a dev environment rather than real-time in a production environment like can be done in PHP.

Even so,  it would be great to have a core application that supported hooks and modules that were self-contained with Controllers, views, migrations, models, and routes and which the main application could invoke based on hooks.

1. Does anyone know if there is an example of this already being done?
2. Can anyone provide any advice on how you might make AdonisJS modular? i.e. routes outside of start/routes.ts
3. Is it even possible to do any of this in AdonisJS or would I be better starting from scratch without a framework like AdonisJS?

**Comment by zmaglica** [comment]
I think that it is possible. I am from the PHP world too and I do understand what you need. You can use Service Providers with adonis IoC to load any third-party module. Currently, a colleague and I are building our APP using DDD where we separate controllers, routes, ... in different folders and it is working great. 

My suggestion is to wait a bit more to allow maintainers to implement all features and after that, you can start with your experiment.

**Reply by simonjcarr** [reply]
@zmaglica Thanks for sharing your experience. Can you confirm what stack you and your colleague used to build your modular app?

**Reply by zmaglica** [reply]
For now, we are building backend only and it is made 100% in adonis v5. Our goal is to replace each domain to a separate server and build a microservice out of it. What we are currently doing is to create multiple folders inside the root "Domains" folder where we put all logic (routes, middlewares, controllers, views, models, ...). Connecting to adonis is easy. We simply load routes file in `routes.ts` file inside `start` folder.

Also, one guy on the discord channel asked the same thing like you. You can check answer of maintainer here: https://discord.com/channels/423256550564691970/423256550564691972/737607488081494107

**Reply by simonjcarr** [reply]
I have created the following structure in my AdonisJS 5 Project

```
/app
--Modules
----admin
------Controllers
------Routes
------Views
```
I have managed to import routes from my custom Routes folder and I can call a method from my route in my custom Controller
```
import Route from '@ioc:Adonis/Core/Route'
export default () => {
  return Route.group(() => {
    Route.get('/', '/App/Modules/admin/Controllers/AdminController.index')
  }).prefix('/admin')
}
```
The problem I am having now is rendering views from my controller.

```
import { HttpContextContract } from '@ioc:Adonis/Core/HttpContext'

export default class AdminController {
  public index ({ view }: HttpContextContract){
    view.render('/App/Modules/admin/Views/index')
  }
}
```
AdonisJS 5 cannot see the route to the edge template
```
Cannot resolve "/App/Modules/admin/Views/index". Make sure the file exists
```

Any ideas?

**Reply by zmaglica** [reply]
I had the same issue with views. When you run `node ace build`, build script deletes edge files that are outside of the default view folder. You have to place view files inside resource folder

Also, you can use route namespaces to avoid this: `/App/Modules/admin/Controllers/AdminController`

example:

```
Route.group(() => {
  Route.post('/', 'UsersController.store').as('users.store')
}).prefix('/users').namespace("App/User/Controllers/Http")
```

**Reply by simonjcarr** [reply]
@zmaglica That's a bit of bugger. You can build a module that includes everything except views? I wonder if @thetutlage could advise if there is a way around this. 

If it's the only thing stopping a truly modular approach, perhaps it can be looked at as part of the "the people would like" for AdonisJS 5 features.

Thanks for the tip on namespacing.

**Reply by zmaglica** [reply]
@simonjcarr There is a workaround (the ugly one) for the problem with views where you can create a command inside module where you copy view fields to the default adonis resource folder

**Reply by thetutlage** [reply]
It is possible to register views from outside by defining custom disks. On this entire topic, I will rather wait and write proper docs, since it is not something I can explain a reply or two.

Also, once I will start with the documentation, I may find some rough edge, so I will work on them at that time

**Reply by simonjcarr** [reply]
Thank you
